llvm/labs/lab2/sredneva_anastasiya/Loop_Sredneva_A.cpp

```diff
@@ -0,0 +1,86 @@
+#include "llvm/Analysis/LoopInfo.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/Pass.h"
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Passes/PassPlugin.h"
+#include "llvm/Support/raw_ostream.h"
+
+using namespace llvm;
+
+namespace {
+struct LoopWrapperPass : public PassInfoMixin<LoopWrapperPass> {
+  PreservedAnalyses run(Function &F, FunctionAnalysisManager &FAM) {
+    LoopInfo &LI = FAM.getResult<LoopAnalysis>(F);
+    FunctionCallee loopStartFunc = F.getParent()->getOrInsertFunction(
+        "loop_start", FunctionType::getVoidTy(F.getContext()));
+    FunctionCallee loopEndFunc = F.getParent()->getOrInsertFunction(
+        "loop_end", FunctionType::getVoidTy(F.getContext()));
+    for (auto &BB : F) {
+      if (Loop *L = LI.getLoopFor(&BB)) {
+        if (L->getHeader() == &BB) {
+          BasicBlock *Preheader = L->getLoopPreheader();
```

Comment:
If getLoopPreheader returns nullptr, then it means that there are more than one preheader. Creating another one preheader basic block may lead to detecting not a very beginning of a loop.